src/mono/wasi/mono-wasi-driver/driver.c

```diff
@@ -0,0 +1,624 @@
+	assemblies = entry;
+	++assembly_count;
+	return mono_has_pdb_checksum ((char*)data, size);
+}
+
+typedef struct WasmSatelliteAssembly_ WasmSatelliteAssembly;
+
+struct WasmSatelliteAssembly_ {
+	MonoBundledSatelliteAssembly *assembly;
+	WasmSatelliteAssembly *next;
+};
+
+static WasmSatelliteAssembly *satellite_assemblies;
+static int satellite_assembly_count;
+
+void
+mono_wasm_add_satellite_assembly (const char *name, const char *culture, const unsigned char *data, unsigned int size)
+{
+	WasmSatelliteAssembly *entry = g_new0 (WasmSatelliteAssembly, 1);
+	entry->assembly = mono_create_new_bundled_satellite_assembly (name, culture, data, size);
+	entry->next = satellite_assemblies;
+	satellite_assemblies = entry;
+	++satellite_assembly_count;
+}
+
+void
+mono_wasm_setenv (const char *name, const char *value)
+{
+	monoeg_g_setenv (strdup (name), strdup (value), 1);
+}
+
+static void *sysglobal_native_handle;
+int32_t SystemNative_LChflagsCanSetHiddenFlag(void);
+char* SystemNative_GetEnv(char* name);
+intptr_t SystemNative_Dup(intptr_t oldfd);
+int32_t SystemNative_Write(intptr_t fd, const void* buffer, int32_t bufferSize);
+int64_t SystemNative_GetSystemTimeAsTicks();
+int32_t SystemNative_LStat(const char* path, int output);
+int32_t SystemNative_ConvertErrorPlatformToPal(int32_t platformErrno);
+void* SystemNative_LowLevelMonitor_Create();
+void SystemNative_LowLevelMonitor_Acquire(void* monitor);
+void SystemNative_LowLevelMonitor_Release(void* monitor);
+int32_t SystemNative_LowLevelMonitor_TimedWait(void *monitor, int32_t timeoutMilliseconds);
+void SystemNative_LowLevelMonitor_Wait(void* monitor);
+int SystemNative_GetErrNo();
+void SystemNative_SetErrNo(int value);
+
+int32_t SystemNative_Write2(intptr_t fd, const void* buffer, int32_t bufferSize) {
+	// Not sure why, but am getting fd=-1 when trying to write to stdout (which fails), so here's a workaround
+	return SystemNative_Write((int)fd == -1 ? 1: fd, buffer, bufferSize);
+}
+
+static PinvokeImport SystemNativeImports [] = {
+	{"SystemNative_GetEnv", SystemNative_GetEnv },
+	{"SystemNative_LChflagsCanSetHiddenFlag", SystemNative_LChflagsCanSetHiddenFlag },
+	{"SystemNative_Dup", SystemNative_Dup},
+	{"SystemNative_Write", SystemNative_Write2},
+	{"SystemNative_GetSystemTimeAsTicks", SystemNative_GetSystemTimeAsTicks},
+	{"SystemNative_LStat", SystemNative_LStat},
+	{"SystemNative_ConvertErrorPlatformToPal", SystemNative_ConvertErrorPlatformToPal},
+	{"SystemNative_LowLevelMonitor_Create", SystemNative_LowLevelMonitor_Create},
+	{"SystemNative_LowLevelMonitor_Acquire", SystemNative_LowLevelMonitor_Acquire},
+	{"SystemNative_LowLevelMonitor_Release", SystemNative_LowLevelMonitor_Release},
+	{"SystemNative_LowLevelMonitor_TimedWait", SystemNative_LowLevelMonitor_TimedWait},
+	{"SystemNative_LowLevelMonitor_Wait", SystemNative_LowLevelMonitor_Wait},
+	{"SystemNative_GetErrNo", SystemNative_GetErrNo},
+	{"SystemNative_SetErrNo", SystemNative_SetErrNo},
+	{NULL, NULL}
+};
```

Comment:
Of course, we should autogenerate a complete set of pinvoke mappings. This small set is sufficient for the console and webserver samples I've been experimenting with.